File: SeparatingAxisTest/SeparatingAxisTest.m
//
//  SeparatingAxisTest.m
//  SeparatingAxisTest
//
//  Created by Najati Imam on 7/26/12.
//  Copyright (c) 2012 Cyrus Innovation. All rights reserved.
//

#import "SeparatingAxisTest.h"

#import "Range.h"

@implementation SeparatingAxisTest

@synthesize result;

-(void) considerAxis:(Vector*) axis {
    Range* aProjection = [a projectOnto:axis];
    Range* bProjection = [b projectOnto:axis];

    Overlap *overlap = [aProjection overlapWith: bProjection];
    
    if (![overlap overlaps]) {
        separated |= true;
    } else {
        float lengthSquaredOfOverlapAlongThisAxis = [overlap correction]*[overlap correction];
        
        if(lengthSquaredOfOverlapAlongThisAxis < minimumSeparationSquared) {
            minimumSeparationSquared = lengthSquaredOfOverlapAlongThisAxis;
            
            [penetration copyFrom:axis];
            [penetration normalize];
            [penetration scaleBy:[overlap correction]];
        }
    }
}

-(void) computeNormals: (Polygon *) polygon {
    Vector *start = [[polygon points] objectAtIndex:0];
    for (int i = 0; i < [[polygon points] count]; i++) {
        Vector *end = [[polygon points] objectAtIndex:i];
        Vector *perpendicular = [end minus: start];
        [perpendicular normalize];
        [perpendicular flop];
        
        [self considerAxis:perpendicular];
        
        start = end;
    }
}

-(id) initWith:(Polygon *) _a and: (Polygon *) _b {
    self = [super init];
    if (!self) return nil;
        
    a = _a;
    b = _b;
    
    separated = false;
    penetration = [[Vector alloc] init];
    
    minimumSeparationSquared = FLT_MAX;
    
    [self computeNormals: a];
    [self computeNormals: b];
    
    result = [[SATResult alloc] initWithPenetration:penetration andSeparated:separated];
    
    return self;
}

@end

File: SeparatingAxisTest/SeparatingAxisTest.c
//
//  SeparatingAxisTest.m
//  SeparatingAxisTest
//
//  Created by Najati Imam on 7/26/12.
//  Copyright (c) 2012 Cyrus Innovation. All rights reserved.
//

#import "SeparatingAxisTest.h"

#import "Range.h"

#import <float.h>

void consider_axis(SeparatingAxisTestState* state, Vector axis) {
    Range aProjection = project_polygon(state->a, axis);
    Range bProjection = project_polygon(state->b, axis);

    Overlap overlap = create_overlap(aProjection, bProjection);
    
    if (!overlap.overlaps) {
        state->separated |= true;
    } else {
        float lengthSquaredOfOverlapAlongThisAxis = overlap.correction*overlap.correction;
        
        if(lengthSquaredOfOverlapAlongThisAxis < state->minimumSeparationSquared) {
            state->minimumSeparationSquared = lengthSquaredOfOverlapAlongThisAxis;

            state->penetration = axis;
            vector_normalize(&state->penetration);
            vector_scale(&state->penetration, overlap.correction);
        }
    }
}

void compute_normals(SeparatingAxisTestState *state, Polygon polygon) {
    Vector start = polygon.points[polygon.point_count - 1];
    for (int i = 0; i < polygon.point_count; i++) {
        Vector perpendicular = vector_subtract(polygon.points[i], start);
        vector_normalize(&perpendicular);
        vector_flop(&perpendicular);
        
        consider_axis(state, perpendicular);
        
        start = polygon.points[i];
    }
}

SATResult sat_test(Polygon a, Polygon b) {
    SeparatingAxisTestState state;

    state.a = a;
    state.b = b;
        
    state.separated = false;
    
    state.minimumSeparationSquared = FLT_MAX;
    
    compute_normals(&state, a);
    compute_normals(&state, b);

    return sat_result_for(state.penetration, state.separated);
}

